File: data/derived_cot/rq3_prompting/results/translation/r1/cpp_to_py/code_output/MahjongConnect.py
import random
from typing import List, Tuple

class MahjongConnect:
    def __init__(self, board_size: Tuple[int, int], icons: List[str]):
        self.BOARD_SIZE = board_size
        self.ICONS = icons
        self.board = self.create_board()

    def create_board(self) -> List[List[str]]:
        rows, cols = self.BOARD_SIZE
        board = [['' for _ in range(cols)] for _ in range(rows)]
        for i in range(rows):
            for j in range(cols):
                board[i][j] = random.choice(self.ICONS)
        return board

    def is_valid_move(self, pos1: Tuple[int, int], pos2: Tuple[int, int]) -> bool:
        x1, y1 = pos1
        x2, y2 = pos2
        rows, cols = self.BOARD_SIZE

        # check bounds
        if not (0 <= x1 < rows and 0 <= y1 < cols and 0 <= x2 < rows and 0 <= y2 < cols):
            return False
        # non‑same position
        if pos1 == pos2:
            return False
        # must have same icon
        if self.board[x1][y1] != self.board[x2][y2]:
            return False
        # check path
        return self.has_path(pos1, pos2)

    def has_path(self, pos1: Tuple[int, int], pos2: Tuple[int, int]) -> bool:
        visited = set()
        stack = [pos1]

        while stack:
            cur = stack.pop()
            if cur == pos2:
                return True
            if cur in visited:
                continue
            visited.add(cur)
            x, y = cur
            # four directions
            for dx, dy in ((0, 1), (0, -1), (1, 0), (-1, 0)):
                nx, ny = x + dx, y + dy
                if 0 <= nx < self.BOARD_SIZE[0] and 0 <= ny < self.BOARD_SIZE[1]:
                    if (nx, ny) not in visited and self.board[nx][ny] == self.board[x][y]:
                        stack.append((nx, ny))
        return False

    def remove_icons(self, pos1: Tuple[int, int], pos2: Tuple[int, int]) -> None:
        self.board[pos1[0]][pos1[1]] = ' '
        self.board[pos2[0]][pos2[1]] = ' '

    def is_game_over(self) -> bool:
        for row in self.board:
            for icon in row:
                if icon != ' ':
                    return False
        return True

File: data/derived_cot/rq1_traces/translation/glm5.1/py_to_cpp/code_output/MahjongConnect.cpp
#include <vector>
#include <string>
#include <utility>
#include <set>
#include <cstdlib>

class MahjongConnect {
public:
    std::vector<int> BOARD_SIZE;
    std::vector<std::string> ICONS;
    std::vector<std::vector<std::string>> board;

    MahjongConnect(std::vector<int> BOARD_SIZE, std::vector<std::string> ICONS)
        : BOARD_SIZE(std::move(BOARD_SIZE)), ICONS(std::move(ICONS)) {
        board = create_board();
    }

    std::vector<std::vector<std::string>> create_board() {
        std::vector<std::vector<std::string>> b(BOARD_SIZE[0], std::vector<std::string>(BOARD_SIZE[1]));
        for (int i = 0; i < BOARD_SIZE[0]; ++i) {
            for (int j = 0; j < BOARD_SIZE[1]; ++j) {
                b[i][j] = ICONS[std::rand() % ICONS.size()];
            }
        }
        return b;
    }

    bool is_valid_move(std::pair<int, int> pos1, std::pair<int, int> pos2) {
        int x1 = pos1.first, y1 = pos1.second;
        int x2 = pos2.first, y2 = pos2.second;

        // Check if positions are within the game board range
        if (!(0 <= x1 && x1 < BOARD_SIZE[0] && 0 <= y1 && y1 < BOARD_SIZE[1] &&
              0 <= x2 && x2 < BOARD_SIZE[0] && 0 <= y2 && y2 < BOARD_SIZE[1])) {
            return false;
        }

        // Check if the two positions are the same
        if (pos1 == pos2) {
            return false;
        }

        // Check if the two positions have the same icon
        if (board[x1][y1] != board[x2][y2]) {
            return false;
        }

        // Check if there is a valid path between the two positions
        if (!has_path(pos1, pos2)) {
            return false;
        }

        return true;
    }

    bool has_path(std::pair<int, int> pos1, std::pair<int, int> pos2) {
        std::set<std::pair<int, int>> visited;
        std::vector<std::pair<int, int>> stack;
        stack.push_back(pos1);

        while (!stack.empty()) {
            std::pair<int, int> current_pos = stack.back();
            stack.pop_back();

            if (current_pos == pos2) {
                return true;
            }

            if (visited.count(current_pos)) {
                continue;
            }

            visited.insert(current_pos);
            int x = current_pos.first;
            int y = current_pos.second;

            // Check adjacent positions (up, down, left, right)
            std::vector<std::pair<int, int>> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
            for (const auto& dir : directions) {
                int new_x = x + dir.first;
                int new_y = y + dir.second;
                if (0 <= new_x && new_x < BOARD_SIZE[0] && 0 <= new_y && new_y < BOARD_SIZE[1]) {
                    if (!visited.count({new_x, new_y}) && board[new_x][new_y] == board[x][y]) {
                        stack.push_back({new_x, new_y});
                    }
                }
            }
        }

        return false;
    }

    void remove_icons(std::pair<int, int> pos1, std::pair<int, int> pos2) {
        int x1 = pos1.first, y1 = pos1.second;
        int x2 = pos2.first, y2 = pos2.second;
        board[x1][y1] = " ";
        board[x2][y2] = " ";
    }

    bool is_game_over() {
        for (const auto& row : board) {
            for (const auto& icon : row) {
                if (icon != " ") {
                    return false;
                }
            }
        }
        return true;
    }
};